tests/benchdnn/dnnl_memory.cpp

```diff
@@ -548,6 +558,72 @@ void dnn_mem_t::memset(int value, size_t size) const {
 }
 
+dnnl_status_t dnn_mem_t::memset_rng(size_t size) const {
+    bool is_opencl = is_opencl_engine(engine_);
+    bool is_sycl = is_sycl_engine(engine_);
+    auto mem = m_padded_ ? m_padded_ : m_;
+    void *mem_handle;
+    DNN_SAFE_V(dnnl_memory_get_data_handle(mem, &mem_handle));
+    if (is_opencl) {
+#if DNNL_GPU_RUNTIME == DNNL_RUNTIME_OCL
+        stream_t stream(engine_);
+        cl_command_queue ocl_queue;
+        cl_context ocl_ctx;
+        cl_device_id ocl_device;
+        cl_program philox_program;
+        cl_kernel ocl_kernel;
+        cl_int err; 
+        DNN_SAFE_V(dnnl_ocl_interop_stream_get_command_queue(stream, &ocl_queue));
+        DNN_SAFE_V(dnnl_ocl_interop_engine_get_context(engine_, &ocl_ctx));
+        DNN_SAFE_V(dnnl_ocl_interop_get_device(engine_, &ocl_device));
+        philox_program = clCreateProgramWithSource(ocl_ctx, 1, &philox_rng_source, nullptr, &err);
+        OCL_TEST(clBuildProgram(philox_program, 1, &ocl_device, nullptr, nullptr, nullptr));
+        ocl_kernel = clCreateKernel(philox_program, "philox_fill_kernel", &err);
+
+        //TODO: generate random seed to avoid repeating values
+        uint64_t seed = 12345;
+        const uint64_t buffSize = (uint64_t)size;
+        const uint64_t blockSize = fmax(16, size / 65536);
+        const size_t gws = size / blockSize + (size % blockSize > 0 ? 1 : 0);
+        switch (memory_kind) {
+            case memory_kind_ext_t::buffer: {
+                auto buf = static_cast<cl_mem>(mem_handle);
+                clSetKernelArg(ocl_kernel, 0, sizeof(cl_mem), &buf);
+                clSetKernelArg(ocl_kernel, 1, sizeof(uint64_t), &buffSize);
+                clSetKernelArg(ocl_kernel, 2, sizeof(uint64_t), &blockSize);
+                clSetKernelArg(ocl_kernel, 3, sizeof(uint64_t), &seed);
+                clEnqueueNDRangeKernel(ocl_queue, ocl_kernel,
+                                    1, nullptr, &gws, nullptr,
+                                    0, nullptr, nullptr);
+
+                DNN_SAFE_V(dnnl_stream_wait(stream));
+                return dnnl_status_t::dnnl_success;
+            }
+            case memory_kind_ext_t::usm:
+            case memory_kind_ext_t::usm_device:
+            case memory_kind_ext_t::usm_shared: {
+                DNN_SAFE_V(dnnl::impl::xpu::ocl::usm::set_kernel_arg(engine_, ocl_kernel, 0, mem_handle));
+                clSetKernelArg(ocl_kernel, 1, sizeof(uint64_t), &buffSize);
+                clSetKernelArg(ocl_kernel, 2, sizeof(uint64_t), &blockSize);
+                clSetKernelArg(ocl_kernel, 3, sizeof(uint64_t), &seed);
+
+                clEnqueueNDRangeKernel(ocl_queue, ocl_kernel, 1, nullptr, &gws, nullptr, 0, nullptr, nullptr);
+                DNN_SAFE_V(dnnl_stream_wait(stream));
+                return dnnl_status_t::dnnl_success;
+            }
+        }
+#endif
+    } else if (is_sycl) {
+#ifdef DNNL_WITH_SYCL
+        // TODO: add sycl support
+         return dnnl_status_t::dnnl_unimplemented;
```

Comment:
There should be some fixed size window where data compression can happen, and I believe it is on the order of 256 bytes, so it is not particularly big. What if we (ab)use the binary primitive to broadcast random data into the buffer by doing `binary_max(uncompressed_src, uncompressed_src)` with broadcast semantics? That should be a bit simpler as then we won't have to maintain all this runtime specific code. although it will introduce memory size alignment requirements.

Reply:
I don't follow - how does `binary_max` generate random data?

Reply:
Binary max doesn't generate the data, instead we setup 1 incompressible source buffer at startup. Essentially, we do something like
```C++
out_data get_mem_rng(size) {

    static dnnl_mem isrc = []() {
        uint8_t rand_data[256];
        fill_rng(rand_data);
        dnnl_mem isrc;
        reorder(isrc, rand_data)
        return isrc;
    }

    dnnl_mem out_data(utils::rnd_up(size, 256))
    out_data = binary_max(isrc, isrc);
    return out_data;
}
```

Reply:
@rjoursler Do you have details on how compression is done? For example, if it can detect a repeating pattern at the page granularity (4 kilobytes) then any pattern of a power of two size would result in some subset of pages having the same data which exposes compression opportunities for the hardware.

Reply:
Discussed with Roy offline - the current hardware seems to work with 256-byte blocks (and future hardware may have possibly larger blocks). So using a small block of 256 bytes with random data as a pattern should be enough. 

But I'm not sure about this trick with broadcasting via binary primitive as it looks like `dst` must match either `src0` or `src1`:

https://github.com/oneapi-src/oneDNN/blob/a6a303a0805687f965453bbf7b66eb34cb6ebe6f/src/common/binary.cpp#L130-L131

Reply:
Thanks for finding that @echeresh, it sounds like a separate kernel would would still be required.